crates/iceberg/src/spec/manifest.rs

```diff
@@ -0,0 +1,671 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Manifest for Iceberg.
+use super::{FormatVersion, ManifestContentType, PartitionSpec, Schema, Struct};
+use super::{Literal, Type};
+use crate::{Error, ErrorKind};
+use apache_avro::{from_value, Reader as AvroReader, Schema as AvroSchema};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+/// A manifest contains metadata and a list of entries.
+pub struct Manifest {
+    metadata: ManifestMetadata,
+    entries: Vec<ManifestEntry>,
+}
+
+impl Manifest {
+    /// Parse manifest from bytes of avro file.
+    pub fn parse_avro(bs: &[u8]) -> Result<Self, Error> {
+        let reader = AvroReader::new(bs)?;
+
+        // Parse manifest metadata
+        let meta = reader.user_metadata();
+        let metadata = ManifestMetadata::parse(meta)?;
+
+        // Parse manifest entries
+        let partition_type =
+            Type::Struct(metadata.partition_spec.partition_type(&metadata.schema)?);
+        let mut entries = Vec::<ManifestEntry>::new();
+
+        match metadata.format_version {
+            FormatVersion::V1 => {
+                let reader = AvroReader::with_schema(Self::v1_schema(), bs)?;
+                for value in reader {
+                    entries.push(
+                        from_value::<_serde::ManifestEntryV1>(&value?)?
+                            .try_into(&partition_type, &metadata.schema)?,
+                    );
+                }
+            }
+            FormatVersion::V2 => {
+                let reader = AvroReader::with_schema(Self::v2_schema(), bs)?;
+                for value in reader {
+                    entries.push(
+                        from_value::<_serde::ManifestEntryV2>(&value?)?
+                            .try_into(&partition_type, &metadata.schema)?,
+                    );
+                }
+            }
+        };
```

Comment:
Would it be possible to iterate over the `reader` values, map the values to `ManifestEntry` and then collect the iterator combinator into the `entries` vec. This way you might save on reallocations of the `entries` vector.

Reply:
Yes!